autoPyTorch/data/utils.py

```diff
@@ -0,0 +1,309 @@
+        np.float32: np.float32,
+        np.float64: np.float32,
+        np.int32: np.int32,
+        np.int64: np.int32
+    }
+
+    # In spite of the names, np.float96 and np.float128
+    # provide only as much precision as np.longdouble,
+    # that is, 80 bits on most x86 machines and 64 bits
+    # in standard Windows builds.
+    if hasattr(np, 'float96'):
+        _mapping[np.float96] = np.float64
+
+    if hasattr(np, 'float128'):
+        _mapping[np.float128] = np.float64
+
+    @classmethod
+    def __getitem__(cls, item: type) -> type:
+        for k, v in cls._mapping.items():
+            if k == item:
+                return v
+        raise KeyError(item)
+
+    @classmethod
+    def __iter__(cls) -> Iterator[type]:
+        return iter(cls._mapping.keys())
+
+    @classmethod
+    def __len__(cls) -> int:
+        return len(cls._mapping)
+
+
+reduction_mapping = _DtypeReductionMapping()
+supported_precision_reductions = list(reduction_mapping)
+
+
+def reduce_precision(
+    X: DatasetCompressionInputType
+) -> Tuple[DatasetCompressionInputType, DatasetDTypeContainerType, DatasetDTypeContainerType]:
+    """ Reduces the precision of a dataset containing floats or ints
+
+    Note:
+        For dataframe, the column's precision is reduced using pd.to_numeric.
+
+    Args:
+        X:  DatasetCompressionInputType
+            The data to reduce precision of.
+
+    Returns:
+        Tuple[DatasetCompressionInputType, DatasetDTypeContainerType, DatasetDTypeContainerType]
+            Returns the reduced data X along with the dtypes it and the dtypes it was reduced to.
+    """
+    precision: Optional[DatasetDTypeContainerType] = None
+    if isinstance(X, np.ndarray) or issparse(X):
+        dtypes = X.dtype
+        if X.dtype not in supported_precision_reductions:
+            raise ValueError(f"X.dtype = {X.dtype} not equal to any supported"
+                             f" {supported_precision_reductions}")
+        precision = reduction_mapping[X.dtype]
+        X = X.astype(precision)
+    elif hasattr(X, 'iloc'):
+        dtypes = dict(X.dtypes)
+
+        integer_columns = []
+        float_columns = []
+
+        for col, dtype in dtypes.items():
+            if is_numeric_dtype(dtype):
+                if is_float_dtype(dtype):
+                    float_columns.append(col)
+                else:
+                    integer_columns.append(col)
+
+        if len(integer_columns) > 0:
+            X[integer_columns] = X[integer_columns].apply(lambda column: pd.to_numeric(column, downcast='integer'))
+        if len(float_columns) > 0:
+            X[float_columns] = X[float_columns].apply(lambda column: pd.to_numeric(column, downcast='float'))
+        precision = dict(X.dtypes)
+    else:
+        raise ValueError(f"Unrecognised data type of X, expected data type to "
+                         f"be in (np.ndarray, spmatrix, pd.DataFrame), but got :{type(X)}")
+
+    return X, precision, dtypes
```

Comment:
The variable names are confusing?
It seems both precision and dtypes are precision and, at the same time, dtypes.
THe difference seems to be before or after, so we should rename them more informative...

Reply:
sure